src/lib/updating-element.ts

```diff
@@ -0,0 +1,545 @@
+  }
+
+  private _validationState: ValidationState = disabled;
+  private _isReflectingProperty: boolean = false;
+  private _instanceProperties: PropertyValues|undefined;
+  private _validatePromise: any = undefined;
+  private _validateResolver: (() => void)|undefined;
+  private _firstUpdated: boolean = false;
+
+  /**
+   * Object with keys for all properties with their current values.
+   */
+  private _props: PropertyValues = {};
+
+  /**
+   * Object with keys for any properties that have changed since the last
+   * update cycle with previous values.
+   */
+  private _changedProps?: PropertyValues|undefined;
+
+  /**
+   * Node or ShadowRoot into which element DOM should be rendered. Defaults
+   * to an open shadowRoot.
+   */
+  protected renderRoot?: Element|DocumentFragment;
+
+  constructor() {
+    super();
+    this.initialize();
+  }
+
+  /**
+   * Performs element initialization. By default this calls `createRoot` to
+   * create the element `root` node and captures any pre-set values for
+   * registered properties.
+   */
+  initialize() {
+    this.renderRoot = this.createRenderRoot();
+    // Apply any properties set on the instance before upgrade time.
+    for (const p in (this.constructor as typeof UpdatingElement)._classProperties) {
+      if (this.hasOwnProperty(p)) {
+        const value = this[p as keyof this];
+        delete this[p as keyof this];
+        this._instanceProperties = this._instanceProperties || {};
+        // NOTE: must capture these into a bag and reset at when validating
+        // to avoid stomping on a user value set in the constructor. Being
+        // async doesn't help here since the subclass' constructor value should
+        // be overwritten.
+        this._instanceProperties![p] = value;
+      }
+    }
+  }
+
+  /**
+   * Returns the node into which the element should render and by default
+   * creates and returns an open shadowRoot. Implement to customize where the
+   * element's DOM is rendered. For example, to render into the element's
+   * childNodes, return `this`.
+   * @returns {Element|DocumentFragment} Returns a node into which to render.
+   */
+  protected createRenderRoot(): Element|ShadowRoot {
+    return this.attachShadow({mode : 'open'});
+  }
+
+  /**
+   * Uses ShadyCSS to keep element DOM updated.
+   */
+  connectedCallback() {
+    if (this._validationState !== disabled && typeof window.ShadyCSS !== 'undefined') {
+      window.ShadyCSS.styleElement(this);
+    }
+    this.invalidate();
+  }
+
+  /**
+   * Synchronizes property values when attributes change.
+   */
+  attributeChangedCallback(name: string, old: string, value: string) {
+    if (old !== value) {
+      this._attributeToProperty(name, value);
+    }
+  }
+
+  /**
+   * Returns the value of the property with `name`. This method is used
+   * to get property values of registered properties.
+   */
+  protected getProperty(name: string) {
```

Comment:
This is an idea for a bigger change that we should probably discuss in person, but...

Could we possibly get rid of `getProperty`, `setProperty` and `_props` altogether?

The way it would work is:

1. `makeProperty` would create a Symbol key for the private storage of the property and defineProperty for the private storage directly on the prototype.
2. The `get` and `set` methods in `makeProperty` would read/write to the private storage property.
3. The setter would do all the `setProperty` work (except write to `_props`), including still writing to `_changedProps`
4. If a user wants to write their own getter/setter pair, they implement their own storage (or call makeProperty), and call invalidate rather than `setProperty`.
5. Any previous call to `setProperty` or `getProperty` should be fine as calls to the accessors.

Reducing metaprograming APIs and forcing properties to be declared will be good for static analysis and compilers. If we remove `getProperty`/`setProperty`, the only remaining reason to disallow class property renaming will be property bindings in lit-html templates, but we could address that with plugins that support renaming in templates.

This should also make getting property values marginally faster due to skipping an object dereference into `this._props`.

WDYT?

Reply:
@justinfagnani we have implemented this on top of lit-html in the past weeks and are very happy with how well it replaces the much bigger metaprogramming API as you call it and how well the developer workflow works. You just assume reactivity works and continue with the class attributes as if they were normal values. It actually feels very nice. We even optionally defined a proxy which allows for deeper observations in objects (performance, I know...). I would share code, but it's deeply entangled with our rxjs integration, so probably not very readable for you guys. Anyways, this is basically just a :+1: :)

Reply:
Done.